backends/dpdk/dpdkAsmOpt.h

```diff
@@ -290,6 +291,246 @@ class ShortenTokenLength : public Transform {
+    }
+
+    bool preorder(const IR::DpdkChecksumSubStatement* c) override {
+        usesInfo[c->field->toString()]++;
+        return false;
+    }
+
+    bool preorder(const IR::DpdkGetHashStatement* c) override {
+        usesInfo[c->dst->toString()]++;
+        return false;
+    }
+
+    bool preorder(const IR::DpdkVerifyStatement* v) override {
+        usesInfo[v->condition->toString()]++;
+        usesInfo[v->error->toString()]++;
+        return false;
+    }
+
+    bool preorder(const IR::DpdkMeterDeclStatement* c) override {
+        usesInfo[c->size->toString()]++;
+        return false;
+    }
+
+    bool preorder(const IR::DpdkMeterExecuteStatement* e) override {
+        usesInfo[e->index->toString()]++;
+        if (e->length)
+            usesInfo[e->length->toString()]++;
+        usesInfo[e->color_in->toString()]++;
+        usesInfo[e->color_out->toString()]++;
+        return false;
+    }
+
+    bool preorder(const IR::DpdkCounterCountStatement* c) override {
+        usesInfo[c->index->toString()]++;
+        if (c->incr)
+            usesInfo[c->incr->toString()]++;
+        return false;
+    }
+
+    bool preorder(const IR::DpdkRegisterDeclStatement* r) override {
+            usesInfo[r->size->toString()]++;
+            return false;
+    }
+
+    bool preorder(const IR::DpdkRegisterReadStatement* r) override {
+            usesInfo[r->index->toString()]++;
+            defInfo[r->dst->toString()]++;
+            return false;
+    }
+
+    bool preorder(const IR::DpdkRegisterWriteStatement* r) override {
+            usesInfo[r->index->toString()]++;
+            return false;
+    }
+
+    bool preorder(const IR::DpdkTable *t) override {
+        auto keys = t->match_keys;
+        if (keys)
+        for (auto ke : keys->keyElements) {
+            dontEliminate.insert(ke->expression->toString());
+        }
+        return false;
+    }
+
+    bool haveSingleUseDef(cstring str) {
+        return defInfo[str] == 1 && usesInfo[str] == 1;
+    }
+
+    void clear() {
+        replacementMap.clear();
+        usesInfo.clear();
+        defInfo.clear();
+    }
+};
+
+// This pass identifies redundant copies/moves and eliminates them.
+class CopyPropagationAndElimination : public Transform {
+    std::unordered_map<cstring, int> newUsesInfo;
+    P4::TypeMap* typeMap;
+    CollectUseDefInfo* collectUseDef;
+
+ public:
+    explicit CopyPropagationAndElimination(P4::TypeMap* typeMap) : typeMap(typeMap) {
+        collectUseDef = nullptr;
+    }
+
+    const IR::Expression* getIrreplaceableExpr(cstring str, bool allowConst);
+    const IR::Expression* replaceIfCopy(const IR::Expression *expr, bool allowConst = true);
+    const IR::DpdkAsmStatement* elimCastOrMov(const IR::DpdkAsmStatement* stmt);
+    IR::IndexedVector<IR::DpdkAsmStatement> copyPropAndDeadCodeElim(
+                                                IR::IndexedVector<IR::DpdkAsmStatement> stmts);
+
+    void clear() {
+        if (collectUseDef)
```

Comment:
another alternative is to allocate a new visitor for each visit instead of reusing it.

Reply:
done